src/mlpack/methods/lsh_model/lshmodel_impl.hpp

```diff
@@ -0,0 +1,505 @@
+    Log::Fatal << "Sampling rate must be floating point number in (0, 1]"
+        << std::endl;
+
+  // Update the object's max K value information.
+  maxKValue = k;
+
+  // Save pointer to training set.
+  this->referenceSet = &referenceSet;
+
+  // Step 1. Select a random sample of the dataset. We will work with only that
+  // sample.
+  arma::vec sampleHelper(referenceSet.n_cols, arma::fill::randu);
+
+  // Keep a sample of the dataset: We have uniformly random numbers in [0, 1],
+  // so we expect about N*sampleRate of them to be in [0, sampleRate).
+  arma::mat sampleSet = referenceSet.cols(
+        arma::find(sampleHelper < sampleRate));
+  // Shuffle to be impartial (in case dataset is sorted in some way).
+  sampleSet = arma::shuffle(sampleSet);
+  const size_t numSamples = sampleSet.n_cols; // Points in sampled set.
+
+  Log::Info << "Training model with " << numSamples << " points in sample set."
+    << std::endl;
+
+  // Step 2. Compute all-vs-all distances of points in the sample.
+  // The distance matrix is symmetric, so we only compute elements above the
+  // diagonal. There are (N * (N - 1)) / 2 such elements.
+  Timer::Start("pairwise_distances");
+  arma::vec distances(numSamples * (numSamples - 1) / 2);
+  size_t d = 0; // Index of where to store next.
+  for (size_t i = 0; i < numSamples; ++i)
+    for (size_t j = i + 1; j < numSamples; ++j)
+      distances(d++) = metric::EuclideanDistance::Evaluate(
+          sampleSet.unsafe_col(i), sampleSet.unsafe_col(j));
+  Log::Info << "Computed " << d << " pointwise distances." << std::endl;
+  Timer::Stop("pairwise_distances");
+
+  // Step 3. Estimate statistics of these distances: log(mean(d)), mean(log(d)),
+  // mean(d).
+  distances = arma::pow(distances, 2);
+  this->meanDist = arma::mean(distances);
+  this->logMeanDist = std::log(meanDist);
+  this->meanLogDist = arma::mean(arma::log(distances));
+
+  // Step 4. Select a small part of the sample as 'anchor points'. Use the rest
+  // of the sample as the reference set. Find the k-Nearest Neighbors' distances
+  // from the anchor points for increasing portion of the reference set. Compute
+  // the arithmetic and geometric mean of distances from each anchor to its
+  // k-Nearest Neighbor.
+  // The geometric mean of N numbers is the Nth root of the product of the
+  // numbers. Through logarithmic properties though, this becomes computable
+  // through exponentiating the mean of the logarithms of x:
+  // exp(mean(log(x))) = geometricmean(x).
+
+  // Number of samples to create for modeling the Gamma Distributions
+  size_t regressionExamples = 50; // TODO: parameter?
+
+  // Number of points to use as queries.
+  size_t numAnchors = (size_t) std::round(0.1 * numSamples);
+  arma::mat queryMat = sampleSet.cols(0, numAnchors - 1);
+  // Evenly spaced sample sizes.
+  arma::Col<size_t> referenceSizes = arma::conv_to< arma::Col<size_t> >::from(
+    arma::linspace(numAnchors, numSamples - numAnchors - 1,
```

Comment:
Can't you call linspace with `arma::Col<size_t>` as a template parameter so you can avoid the `conv_to`?  I think that is possible.